File: Loenn/entities/ParallaxDecal.lua
local drawSprite = require("structs.drawable_sprite")

local ParallaxDecal = {}

ParallaxDecal.name = "OzmasOddities/ParallaxDecal"
ParallaxDecal.placements = {
    name = "ParallaxDecal",
    data = {
        Path = "",
        ScaleX = 1.0,
        ScaleY = 1.0,
        Depth = 9500,
        Rotation = 0,
        Color = "FFFFFFFF",
        ParallaxX = 0.1,
        ParallaxY = 0.1
    }
}

ParallaxDecal.fieldOrder = {"x","y","ParallaxX","ParallaxY","Path","Color","ScaleX","ScaleY","Rotation","Depth"}

function ParallaxDecal.texture(room, entity)
    if drawSprite.fromTexture("decals/" .. entity.Path .. "00") ~= nil then
        return "decals/" .. entity.Path .. "00"
    else
        return "decals/" .. entity.Path
    end
end

function ParallaxDecal.scale(room, entity)
    return { entity.ScaleX or 1, entity.ScaleY or 1 }
end
function ParallaxDecal.rotation(room, entity)
    return (entity.Rotation or 0) * math.pi / 180
end

return ParallaxDecal

File: Source/Entities/ParallaxDecal.cs
using Monocle;
using Microsoft.Xna.Framework;
using Celeste.Mod.Entities;

namespace Celeste.Mod.OzmasOddities.Entities
{
    [CustomEntity("OzmasOddities/ParallaxDecal")]
    public class ParallaxDecal : Decal
    {
        private Vector2 parallaxVector;
        public ParallaxDecal(EntityData data, Vector2 offset)
            : base(data.Attr("Path", ""), data.Position + offset, new Vector2(data.Float("ScaleX", 1f), data.Float("ScaleY", 1f)), data.Int("Depth", 9500))
        {
            Rotation = MathHelper.ToRadians(data.Float("Rotation", 0f));
            Color = Calc.HexToColorWithAlpha(data.Attr("Color", "FFFFFFFF"));
            parallaxVector = new Vector2(data.Float("ParallaxX", 0.1f), data.Float("ParallaxY", 0.1f));
            parallax = false;
            wave = null;
        }

        public override void Render()
        {
            Vector2 pos = Position;
            Camera camera = SceneAs<Level>().Camera;
            Vector2 camOffset = camera.Position + new Vector2(160f, 90f);
            Vector2 parallaxOffset = Vector2.Multiply(parallaxVector, Position - camOffset);
            Position += parallaxOffset;
            base.Render();
            Position = pos;
        }
    }
}
